### ISSUE TITLE
Add support for props destructure

### ISSUE THREAD

**Issue by ota-meshi** [vuejs/eslint-plugin-vue#2121]
**Tell us about your environment**

- **ESLint version:**  v8.36.0
- **eslint-plugin-vue version:** v9.10.0
- **Node version:** v18.12.1

**The problem you want to solve.**

Add support for props destructure. It will be introduced in Vue 3.3.

```vue
<script setup lang="ts">
const { foo = 1, bar = 'ok' } = defineProps<{ foo?: number, bar?: string }>()
</script>
```

**Your take on the correct solution to problem.**

I think we should at least change the rules in the following list.

- [ ] Change `vue/no-setup-props-destructure` rule.
        Destructuring assignment immediately after defineProps() is allowed.
        However, Vue2 doesn't allow to use props destructure, so we may need to split rules.
- [ ] Change `vue/require-valid-default-prop` rule.
- [ ] Change `vue/require-default-prop` rule.
- [ ] Change `vue/no-boolean-default` rule.
        We need to support a new way to specify defaults.

**Additional context**

https://github.com/vuejs/core/pull/7986

**Comment by ota-meshi** [comment]
Since it's still an experimental feature, eslint-plugin-vue won't support it yet until it's GA.

**Comment by Jungzl** [comment]
It is now a [stable feature](https://github.com/vuejs/core/pull/7986#issue-1646843263) I think.

**Comment by n0099** [comment]
It's now stabled in 3.5.0

**Comment by henribru** [comment]
Perhaps a bit niche, but a rule that bans props destructure could also be useful, to keep consistency with existing code that doesn't use it

**Comment by ota-meshi** [comment]
I believe we have merged the necessary changes into the master branch so I will close this issue.
